crates/neon/src/lifecycle.rs

```diff
@@ -54,6 +59,176 @@ pub(crate) struct InstanceData {
+            _ => None,
+        }
+    }
+
+    pub(crate) fn get_or_init<'cx, 'a, C>(
+        cx: &'a mut C,
+        id: usize,
+        value: GlobalCellValue,
+    ) -> &mut GlobalCellValue
+    where
+        C: Context<'cx>,
+    {
+        InstanceData::globals(cx)
+            .get(id)
+            .pre_init(|| GlobalCell::Init(value));
+
+        GlobalCell::get(cx, id).unwrap()
+    }
+
+    pub(crate) fn get_or_init_with<'cx, 'a, C, F>(
+        cx: &'a mut C,
+        id: usize,
+        f: F,
+    ) -> &mut GlobalCellValue
+    where
+        C: Context<'cx>,
+        F: FnOnce() -> GlobalCellValue,
+    {
+        InstanceData::globals(cx)
+            .get(id)
+            .pre_init(|| GlobalCell::Init(f()));
+
+        GlobalCell::get(cx, id).unwrap()
+    }
+
+    pub(crate) fn get_or_try_init<'cx, 'a, C, F>(
+        cx: &'a mut C,
+        id: usize,
+        f: F,
+    ) -> NeonResult<&mut GlobalCellValue>
+    where
+        C: Context<'cx>,
+        F: FnOnce(&mut C) -> NeonResult<GlobalCellValue>,
+    {
+        // Kick off a new transaction and drop it before getting the result.
+        {
+            let mut tx = TryInitTransaction::new(cx, id);
+            tx.run(|cx| Ok(f(cx)?))?;
+        }
+
+        // If we're here, the transaction has succeeded, so get the result.
+        Ok(GlobalCell::get(cx, id).unwrap())
+    }
+}
+
+impl Default for GlobalCell {
+    fn default() -> Self {
+        GlobalCell::Uninit
+    }
+}
+
+impl GlobalTable {
+    pub(crate) fn get(&mut self, index: usize) -> &mut GlobalCell {
+        if index >= self.cells.len() {
+            self.cells.resize_with(index + 1, Default::default);
+        }
+        &mut self.cells[index]
+    }
+}
+
+/// An RAII implementation of `GlobalCell::get_or_try_init`, which ensures that
+/// the state of a cell is properly managed through all possible control paths.
+/// As soon as the transaction begins, the cell is labelled as being in a dirty
+/// state (`GlobalCell::Trying`), so that any additional re-entrant attempts to
+/// initialize the cell will fail fast. The `Drop` implementation ensures that
+/// after the transaction, the cell goes back to a clean state of either
+/// `GlobalCell::Uninit` if it fails or `GlobalCell::Init` if it succeeds.
+struct TryInitTransaction<'cx, 'a, C: Context<'cx>> {
+    cx: &'a mut C,
+    id: usize,
+    _lifetime: PhantomData<&'cx ()>,
+}
+
+impl<'cx, 'a, C: Context<'cx>> TryInitTransaction<'cx, 'a, C> {
+    fn new(cx: &'a mut C, id: usize) -> Self {
+        let mut tx = Self {
+            cx,
+            id,
+            _lifetime: PhantomData,
+        };
+        tx.cell().pre_init(|| GlobalCell::Trying);
+        tx
+    }
+
+    /// _Post-condition:_ If this method returns an `Ok` result, the cell is in the
+    /// `GlobalCell::Init` state.
+    fn run<F>(&mut self, f: F) -> NeonResult<()>
+    where
+        F: FnOnce(&mut C) -> NeonResult<GlobalCellValue>,
+    {
+        if self.is_trying() {
```

Comment:
Shouldn't this panic instead of succeeding?

Reply:
No, the `run` method is the actual body of the transaction. We panic in `new` if we were already `Trying`. Otherwise, `new()` _sets_ the state from `Uninit` to `Trying`. But if we are already initialized there's nothing to do. So this conditional is our way of checking to see if we have work to do or not.

Reply:
Ah! Thanks for explaining.